src/lib/descriptor/mod.rs

```
@@ -14,69 +14,175 @@ mod cargo_alias;
-                        ) => {
-                            let mut base_profile_env_mut = base_profile_env.clone();
-                            let mut extended_profile_env_mut = extended_profile_env.clone();
+        // if the env variable is in the current scope add add an edge,
+        // otherwise it is referencing and external variable.
+        for used in merge_env_depends_on(val).into_iter() {
+            if graph.contains_node(used) {
+                graph.add_edge(*key, used, ());
+            }
+        }
+    }
 
-                            let merged_sub_env =
-                                merge_env(&mut base_profile_env_mut, &mut extended_profile_env_mut);
+    let variables = match toposort(&graph, None) {
+        Ok(iter) => iter,
+        Err(_) => {
+            // cycle has been detected, for better performance we now only
+            // execute scc.
+            // In strongly-connected-components every vertex
+            // (node) is reachable from every other node.
+            // This means that there **must** be a cycle.
+            // This isn't strictly necessary, but aids when debugging.
+            let mut err = "A cycle between different env variables has been detected.".to_owned();
+            for scc in kosaraju_scc(&graph) {
+                let render = scc
+                    .iter()
+                    .chain(scc.first())
+                    .map(ToString::to_string)
+                    .reduce(|acc, name| format!("{acc} -> {name}"));
+
+                if let Some(render) = render {
+                    err.push_str(&format!(" Cycle: {}.", render));
+                }
+            }
 
-                            merged.insert(key_str, EnvValue::Profile(merged_sub_env));
-                        }
-                        _ => {
-                            merged.insert(key_str, value_clone);
-                        }
-                    };
+            return Err(err);
```

breaking the build due to a cycle is a new behavior. I hope this won't break other projects.

---

I don't think so. I believe, in general, this is an improvement because it tells the user specifically why a configuration won't work. Pre PR, this would just silently error out, meaning it would lead to more frustration if that makes sense.

---

agree with approach. just concerned. later, i'm getting lots of issues from unhappy people :)

---

I see. This is why specifically search for strongly connected components so that people can see **where** a cycle is.

So if you declare

```
ENV2 = ${ENV1}
ENV1 = ${ENV2}
```

then it will tell you: cycle detected: ENV2 -> ENV1 -> ENV2.

I think this should help minimize the number of issues. Maybe I can clarify the issue message even more, to redirect to the README for a link explaining how this works.

---

cool